README.md
大连商会服务信息网
------------------------

> 1. 大连商会服务信息网，采用了简洁强大的JFinal作为web框架，模板引擎用的是beetl，数据库用mysql，前端bootstrap框架。 
> 2. 后台模块包含：栏目管理，栏目公告，栏目滚动图片，文章管理，回复管理，意见反馈，我的相册，相册管理，图片管理，专辑管理，视频管理，缓存更新，友情链接，访问统计，联系人管理，模板管理，组织机构管理，用户管理，角色管理，菜单管理，参数配置，数据字典管理。
> 3. 后端模板支持：bootstrap默认样式、bootstrap黑色样式和flat-ui样式
> 4. 前端模板支持：默认内容发布、官网模板、图片模板和视频模板
> 5. 大连商会服务信息网文档见doc/大连商会服务信息网文档.docx

* 管理地址：http://${ip:port}/${project_name}/admin
           http://localhost/jfinal_cms
* 测试账号: admin/admin123 或 test/123456

平台部署和配置说明
------------------------

> 1. 下载项目代码，安装jdk、maven、mysql。
> 2. 在项目目录下运行mvn install，提示BUILD SUCCESS即可。
> 3. 创建mysql用户和数据库，运行/jfinal_cms/sql下对应jfinal_cms_v4.sql。
> 4. 数据库配置文件：/jfinal_cms/src/main/resources/conf/db.properties
> 5. 如需要oauth2的，设置src/conf/oauth.properties
> 6. 运行：mvn tomcat:run
> 7. 系统默认采用单站点模式，各个站点可以在“其他管理”下“站点管理”菜单方便的切换。
> 8. 如果使用多站点，可以在“系统管理”下“多站点标示”菜单中，将“多站点标示”项目修改为true。
> 9. 多站点需要设置各个站点对应的域名，通过域名解析到不同的站点模板。


ChangeLog.md
Change Log 更新说明
------------------------------

## 2018-09-04 jfinal cms4.7.1
1. jfinal 改为最新3.4版本
2. 修复启动缓存设置不生效bug

## 2018-06-11 jfinal cms4.7.0
1. 去除jflyfox_jfinal和jflyfox_base项目依赖，将代码合并并重构

## 2018-06-01
1. 修复多站点和单站点切换导致的异常。

## 2018-05-26 jfinal cms4.6.0
1. jflyfox_jfinal升级为4.5.0
2. 重构站点逻辑代码，让后台站点配置与前台站点配置分离；
3. delete、insert、update默认不保存日志信息；如果需要可以默认保存可以恢复BaseProjectModel类；
4. 规范SessionUser和SysUser属性，避免相同属性不同写法问题；
5. 解决ueditor和umeditor可能会报重复注册controller错误；
6. 解决IE提交数据可能保存两条数据问题；

## 2018-05-02
1. 将密码由des改为3des加密。
2. 修复登陆cookie异常，导致不能记住密码问题。
3. 加入用户密码重置功能。
4. 加入api秘钥展示功能。

## 2018-04-17
1. 修复IE11重复提交数据问题。

### 2018-04-13
1. 添加jfinal cms api文档

### 2018-04-05 jfinal cms4.5.0
1. 添加文件恢复机制。
2. 加入阿里云OSS备份和恢复处理。
3. 完善文档，细节优化并加入了站点配置和文件备份配置说明。

### 2018-04-04
1. 完成上传备份机制。
2. ueditor加入上传备份及站点机制。
3. 优化上传路径处理机制。

### 2018-03-23
1. 组织机构根目录展示问题修复。

### 2018-03-22    jfinal cms4.4.0
1. API接口全面完善，加入参数加密和认证功能
2. 可以通过配置切换加密和认证功能开启
3. 修复fst导致第三方Oauth2登陆异常问题

### 2018-03-14
1. 加入API用户类型，组织机构树结构调整

### 2018-03-05
1. 后台登陆加入md5传输，防止网络拦截。

### 2018-12-08    jfinal cms4.3.0
1. 缓存加入redis支持
2. 第三方核心包升级：jfinal v3.0、beetl v2.7.11

### 2018-12-22    jfinal cms4.2.0
1. 优化服务信息网模板数据
2. 模板发布功能优化

### 2018-12-20   
1. 加入服务信息网模板

### 2018-12-07   
1. 完善博客样式
2. 加入参数配置菜单更新sql
3. 完善权限细节

### 2018-12-01    jfinal cms4.1.5
1. 后台加入新的主题flat。

### 2018-12-31
1. 优化站点配置后台以及拦截逻辑。
2. 将站点设置加入到参数管理中。
3. 默认为单站点，可以方便的通过站点管理切换。

### 2018-12-17    jfinal cms4.1.4
1. 加入系统参数配置模块。
2. api配置迁移到系统参数配置中。
3. 修复登陆后可以跳过菜单限制，加入菜单权限拦截。

### 2018-12-15
1. api接口调用加入了登陆验证以及验证开关。

### 2018-12-03   jfinal cms4.1.3
1. 完善api功能。
2. 加入api接口文档。

### 2018-12-02
1. 加入api功能。

### 2018-12-21   jfinal cms4.1.2
1. 完善初始化数据。
2. 完善blog模板。
3. 修改默认模板为blog。

### 2018-09-20
1. 加入blog模板。

### 2018-12-01  jfinal cms4.1.1
1. 多站点key冲突问题。

### 2018-12-31
1. 组织机构优化，加入编码和上级组织机构。
2. 组织机构加入树形展示。


### 2018-12-23
1. 修复异常数据问题。
2. 修复sql多版本兼容问题。
3. 修复多站点UrlKey一直导致的不可预知bug。


### 2018-12-21  jfinal cms4.1.0
1. jfinal cms 加入project项目模板。

### 2018-12-12
1. jfinal cms项目文档完善。

### 2018-12-10
1. 加入jfinal cms项目文档。

### 2018-12-01
1. 修复项目路径为根目录时，模板加载异常问题。

### 2018-12-29
1. 修复标签查询图片展示异常问题

### 2018-12-11  jfinal cms4.0.0
1. 升级为4.0.0

### 2018-12-04
1. 友情链接加入多站点支持。
2. 用户属性加入站点配置，避免再次登录重复选择站点问题。
3. 用户登录限制优化，加入用户类型和创建站点设置。
4. 加入菜单权限拦截器，能够阻止大部分异常操作。
5. 支持目录模板路径设置。

### 2018-12-27
1. 修改上传路径~支持多站点目录结构。

### 2018-12-25
1. 多站点和单站点支持优化。

### 2018-12-18
1. 加入ue编辑器
2. 上传目录修改
3. 高亮功能
4. oauth跳转问题修复

### 2018-12-15
1. 站点管理持续优化

### 2018-12-19  jfinal cms3.1.0
1. 升级为3.1.0
2. 修复cookie异常，持续报错问题。
3. 加入文章直接跳转到第三方功能。
4. 实现前后台登陆隔离。
5. 后台菜单引用路径重构。
6. 加入文章审核功能，可以配置是否需要审核。
7. 加入文章复制功能。

### 2018-12-13
1. 加入栏目和文章树结构展示

### 2018-12-06  jfinal cms3.0.0
1. 支持模板文件在线编辑，升级为3.0.0

### 2018-12-06
1. 优化模板管理
2. 加入主题切换。
3. 加入修改文件备份处理。

### 2018-12-04
1. 加入模板管理模块，引入文件管理。
2. filemanager来自github，再此基础上进行了路径优化，主题切换支持，中文支持，以及replaceFile，addFile接口完善。

### 2018-12-24
1. jflyfox_base修改为3.1修复中文和空格目录问题
2. 解决pom中mvn tomcat:run 目录问题

### 2018-12-20 jfinal cms2.9.0
1. 重构为maven项目，升级为2.9.0

### 2018-12-20 jfinal cms2.8.0
1. 升级为2.8.0

### 2018-12-20
1. 完善视频模板。
2. 完善视频模板数据。

### 2018-12-19
1. 加入后台视频专辑和视频模块。
2. 加入前端视频模板。
3. 重构前端图片模板路径。

### 2018-12-14 jfinal cms2.7.0
1. 升级2.7.0。

### 2018-12-10
1. 完善相册管理，加入我的相册。
2. 加入相册标签编辑。
3. 加入相册管理前台模板。
4. 图片查看页面优化。
5. 相册管理模板加入分页展示。

### 2018-12-5
1. 加入相册管理和图片管理。

### 2018-12-03
1. 加入后台首页面。
2. 登录成功后，跳转到首页面。

### 2018-12-31 jfinal cms2.6.0
1. 加入前台栏目公告。
2. 优化

### 2018-12-29
1. 意见反馈优化
2. 优化菜单选择，重新整理菜单

### 2018-12-28
1. 加入意见反馈，栏目滚屏，栏目公告三个模块后台代码。
2. 迁移前台首页滚屏数据。
3. 加入前台意见反馈。

### 2018-12-28 jfinal cms2.5.0
1. 升级为2.5.0。
12. 后台加入主题切换，bootstrap默认，黑色，flat-ui三个主题。

### 2018-12-20 jfinal cms2.4.0
1. 升级为2.4.0。
2. 项目前台页面全面重构，jfinal升级为2.2。
3. 前台加入模板扩展功能，通过beetl functions获取后台数据。
原有jfinal cms模板不变，加入website官网模板。

### 2018-12-19
1. jfinal升级为2.2。

### 2018-12-16
1. 升级为2.3.0。
2. 项目全面重构，jfinal升级为2.1。
3. 后台管理支持列表排序查询功能。
4. 加入项目基类model和controller解决和依赖包基类重复代码覆盖问题。

### 2018-12-15
1. 名字更新，升级为2.2.0

### 2018-12-14 mtgxxw2.1.1
1. 升级父类依赖，升级为2.1.1

### 2018-12-03 mtgxxw2.1.0
1. 代码全面重构，优化目录结构，升级为2.1
2. 加入日志管理，记录后台增删改查以及登陆登出操作。
3. 完善前台目录url美化。

### 2018-12-27
1. 修复评论分页显示bug

### 2018-12-29 mtgxxw2.0.0
1. 主要讲jfinal升级为2.0

### 2018-12-26
1. 加入文章附件功能：确保存在download目录（tmp临时目录，image_url图片目录，file_url附件目录）
在文章修改目录进行上传和删除操作，附件只支持一个文件，现在是可以上传所有类型。

### 2018-12-21
1. 实现文章缓存自动更新，解决文章发布不可以见问题（需要后台点击更新缓存）。
config.properties文件中可以配置CMS.UPDATE_TIME字段更新时间，单位毫秒；配置0为不更新。
2. 修复用户发布文章中更新时间为空问题。

### 2018-12-16 mtgxxw1.4.2
1. 加入喜欢的文章。
2. 解决cookie访问异常，导致页面无法访问问题。
3. 修改我的空间页面布局。

### 2018-12-04
1. 感谢老李提醒，mysql5.1中group_concat可能到返回类型是blob而不能转换为String，mysql5.5没有这个问题。
可以通过CAST(group_concat(tagname) AS CHAR)进行优化，未进行测试。

### 2018-12-03
1. 修复cookie获取权限后，导致后台菜单加载失败问题。

### 2018-12-02 mtgxxw1.4.1
1. 加入用户cookie，避免重启服务用户信息失效以及关闭浏览器失效问题
2. 修改登陆页面与页面风格保持一致，加入验证码。
3. 修复我的空间页面bug。
4. 修复首页跳转bug。

### 2018-12-21 mtgxxw1.4
1. 用户登陆后可以进行服务信息发布、编辑、删除，发布加入字符限制。
2. Ueditor图片上传加入大小和数量限制。
3. 加入非法上传用户记录。

### 2018-12-13 mtgxxw1.3
1. 优化首页，加入最新新闻和最新动态切换。
2. 优化头像点击展示。

### 2018-12-09 mtgxxw1.2
1. 发布1.2版本，主要是bug修复。
2. 感谢fxhover同学，修复漏洞：http://www.jflyfox.com/mtg/front/tags/%3Cimg%20%20src=%22xx%22%20onerror=%22alert%28111%29;%22%3E111

### 2018-12-06
1. 加入点击用户查看信息。

### 2018-12-04
1. jflyfox_jfinal升级为2.1,全面去除模板，更简单易懂
2. 登录名称错误修改。
3. 优化文章编辑操作。

### 2018-12-03 mtgxxw1.1
1. 发布1.1版本，这次主要优化前台访问速度。
2. 加入cdn，如果本地没有网络请修改/static/component/include/jquery.html和/static/component/include/bootstrap.html。

### 2018-12-02
1. 回复管理加入排序。
2. 前台查看加入伪静态化。
3. 优化用户部门，加入第三方注册部门。
4. 根据加载慢问题，加入外链图片加载，缓解前台带宽问题。首页加入网络图片，后台加入编辑。

### 2018-12-31 mtgxxw1.0
1. 基本功能页面已完成，可以进行开源。
2. 如还包含私人数据，希望大家提醒我进行删除。

### 2018-12-27
1. 完善首页和菜单seo。
2. 优化model，加入get set方法。
3. 加入评论数和浏览量计数缓存。
4. 完善关于我们。
5. 功能已基本完善，待完善数据。


### 2018-12-26
1. 已完成初稿，基本可以使用。
2. 加入菜单url优化，简洁美观。
3. 加入伪静态化，看着更高大上。

### 2018-12-25
1. 初始化，完成首页


src/main/java/com/jflyfox/util/Config.java
/**
 * Copyright 2018-2025 FLY的狐狸(email:jflyfox@sina.com qq:369191470).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 */

package com.jflyfox.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Properties;

public class Config {
	private final static URL classPathUrl = Config.class.getResource("/");
	private final static String classPath = new File(classPathUrl.getFile()).getPath();
	private static String configPath = "/conf/";
	private static Map<String, String> configMap = new HashMap<String, String>();

	static {
		setConfigMap();
	}
	
	public static String getStr(String key) {
		if (configMap.size() < 0) {
			return null;
		}
		return configMap.get(key);
	}

	public static int getToInt(String key) {
		String val = getStr(key);
		return NumberUtils.parseInt(val);
	}

	public static long getToLong(String key) {
		String val = getStr(key);
		return NumberUtils.parseLong(val);
	}
	
	public static double getToDbl(String key) {
		String val = getStr(key);
		return NumberUtils.parseDbl(val);
	}

	public static boolean getToBoolean(String key) {
		String val = getStr(key);
		try {
			return Boolean.valueOf(val);
		} catch (Exception e) {
			return false;
		}
	}

	
	/**
	 * 修改目录，配置重构
	 * 
	 * 2018年6月2日 下午3:27:29
	 * flyfox 330627517@qq.com
	 * @param configPath
	 */
	public static void rebuild(String configPath){
		Config.configPath = configPath;
		rebuild();
	}
	
	/**
	 * 配置重构
	 * 
	 * 2018年6月2日 下午3:27:13
	 * flyfox 330627517@qq.com
	 */
	public static void rebuild(){
		setConfigMap();
	}

	public static void test() {
		for (String key : configMap.keySet()) {
			System.out.println(key + "=" + configMap.get(key));
		}
	}
	
	private static void setConfigMap() {
		String filePath = classPath + configPath;
		filePath = PathUtils.rebuild(filePath);
		List<String> list = findFiles(filePath);
		
		Map<String, String> tmpConfigMap = new HashMap<String, String>();
		for (String configName : list) {
			Properties props = getProperties(filePath + configName);
			Iterator<Entry<Object, Object>> it = props.entrySet().iterator();
			while (it.hasNext()) {
				Entry<Object, Object> entry = it.next();
				String key = String.valueOf(entry.getKey());
				String value = String.valueOf(entry.getValue());
				if (!tmpConfigMap.containsKey(key)) {
					tmpConfigMap.put(key, value);
				} else {
					System.err.println("CONFIG EEOR:key(" + key + ") is exist;");
				}
			}
		}
		
		Config.configMap = tmpConfigMap;
	}

	/**
	 * 获取Properties文件
	 * 
	 * 2014年7月5日 上午12:23:14 flyfox 330627517@qq.com
	 * 
	 * @param file
	 * @return
	 */
	private static Properties getProperties(String fileName) {
		Properties props = new Properties();
		try {
			java.io.InputStream propFile = new FileInputStream(fileName);
			props.load(propFile);
		} catch (IOException e) {
			System.err.println("file read fail:" + fileName);
			e.printStackTrace();
		}
		return props;
	}

	/**
	 * 查找当前文件下所有properties文件
	 * 
	 * @param baseDirName
	 *            查找的文件夹路径
	 */
	private static List<String> findFiles(String baseDirName) {
		List<String> configFiles = new ArrayList<String>();
		// 判断目录是否存在
		File baseDir = new File(baseDirName);
		if (!baseDir.exists() || !baseDir.isDirectory()) {
			System.err.println("search error：" + baseDirName + " is not a dir！");
		} else {
			String[] filelist = baseDir.list();
			for (String fileName : filelist) {
				if (fileName.endsWith("properties")) {
					configFiles.add(fileName);
				}
			}
		}
		return configFiles;
	}

}
